### PR TITLE
Fix flaky test for testSnapshotRestoreSync

### PR DESCRIPTION
**Pull request by MyEnthusiastic**
## What is the purpose of the change

*Fix flaky test*

## Brief change log

  - *broadcastState2 is based on Map structure.*
  - *When we put two elements into it, before the change, we assume the first entry we get is the <1, 2> and the second entry is <2, 5> which is not necessary right. So the change makes it independent of the order, just check the element inside the Map, which solves the problem*

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

*(Please pick either of the following options)*

  - *This change is to fix the test, it changes the test itself*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no 
  - The runtime per-record code paths (performance sensitive): no 
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable 

Here is the script for running the non-dex

```
mvn -pl flink-runtime -DnondexRuns=10 edu.illinois:nondex-maven-plugin:2.1.7-SNAPSHOT:nondex -Dtest=org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreSync```

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/OperatorStateBackendTest.java [hunk 1]
```diff
@@ -649,20 +649,23 @@ void testSnapshotRestoreSync() throws Exception {
...
+                    throw new Exception("Unexpected key in broadcast state: " + entry.getKey());
+                }
+                count++;
+            }
+            assertThat(count).isEqualTo(2);
 
             bIt = broadcastState2.iterator();
             assertThat(bIt).hasNext();
-            entry = bIt.next();
+            Map.Entry<Serializable, Serializable> entry = bIt.next();
```
Comment (zzjas) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/OperatorStateBackendTest.java]: The original test reuses this variable. Maybe you could also reuse it with the one at Line 654.

**Review by zzjas — COMMENTED** [review]
Left a comment for the code. Other than that, the fix looks ok to me and you can proceed to open a real PR. Once you open a real PR, please mark this tentative PR as `Opened` in your `tentative_pr.csv` file and also raise a PR to IDoFT marking this as `Opened`. Thanks!

**Comment by Suraj-Vashista-BK** [comment]
> ## What is the purpose of the change
> _Fix flaky test_
> 
> ## Brief change log
> * _broadcastState2 is kind of **map** (although I'm not able to find implementation, only python version of implementation in that repository, doubt it is transferred to java as somewhere. I infer that mainly from the put operation and flaky test result )_
> * _we put two elements into it, but when test, we assume the first entry we get is the <1, 2> and the second entry is <2, 5> which is not necessary right._
> 
> ## Verifying this change
> Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing
> 
> _(Please pick either of the following options)_
> 
> * _This change is to fix the test, it changes the test itself_
> 
> ## Does this pull request potentially affect one of the following parts:
> * Dependencies (does it add or upgrade a dependency): no
> * The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
> * The serializers: no
> * The runtime per-record code paths (performance sensitive): no
> * Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
> * The S3 file system connector: no
> 
> ## Documentation
> * Does this pull request introduce a new feature? no
> * If yes, how is the feature documented? not applicable
> 
> Here is the script for running the non-dex
> 
> ```
> mvn -pl flink-runtime -DnondexRuns=10 edu.illinois:nondex-maven-plugin:2.1.7-SNAPSHOT:nondex -Dtest=org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreSync```
> ```

The fix looks okay. One suggestion would be to link you account to the commit. If you check your commit, it does not link to your git account. When you raise a real PR , if your commit is not linked to your git, you wont appear as a contributor.

**Comment by yashdeep97** [comment]
Nice work on the code changes!
Minor comment on the PR description - can you add a short description of the problem and the cause of flakiness in the tests and how the code change addresses the flakiness.

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/OperatorStateBackendTest.java (hunk 1 above)
Comment (MyEnthusiastic) [reply flink-runtime/src/test/java/org/apache/flink/runtime/state/OperatorStateBackendTest.java]: Resolved : Already made that resuable

**Comment by MyEnthusiastic** [comment]
> > ## What is the purpose of the change
> > _Fix flaky test_
> > ## Brief change log
> > 
> > * _broadcastState2 is kind of **map** (although I'm not able to find implementation, only python version of implementation in that repository, doubt it is transferred to java as somewhere. I infer that mainly from the put operation and flaky test result )_
> > * _we put two elements into it, but when test, we assume the first entry we get is the <1, 2> and the second entry is <2, 5> which is not necessary right._
> > 
> > ## Verifying this change
> > Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing
> > _(Please pick either of the following options)_
> > 
> > * _This change is to fix the test, it changes the test itself_
> > 
> > ## Does this pull request potentially affect one of the following parts:
> > 
> > * Dependencies (does it add or upgrade a dependency): no
> > * The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
> > * The serializers: no
> > * The runtime per-record code paths (performance sensitive): no
> > * Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
> > * The S3 file system connector: no
> > 
> > ## Documentation
> > 
> > * Does this pull request introduce a new feature? no
> > * If yes, how is the feature documented? not applicable
> > 
> > Here is the script for running the non-dex
> > ```
> > mvn -pl flink-runtime -DnondexRuns=10 edu.illinois:nondex-maven-plugin:2.1.7-SNAPSHOT:nondex -Dtest=org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreSync```
> > ```
> 
> The fix looks okay. One suggestion would be to link you account to the commit. If you check your commit, it does not link to your git account. When you raise a real PR , if your commit is not linked to your git, you wont appear as a contributor.



> > ## What is the purpose of the change
> > _Fix flaky test_
> > ## Brief change log
> > 
> > * _broadcastState2 is kind of **map** (although I'm not able to find implementation, only python version of implementation in that repository, doubt it is transferred to java as somewhere. I infer that mainly from the put operation and flaky test result )_
> > * _we put two elements into it, but when test, we assume the first entry we get is the <1, 2> and the second entry is <2, 5> which is not necessary right._
> > 
> > ## Verifying this change
> > Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing
> > _(Please pick either of the following options)_
> > 
> > * _This change is to fix the test, it changes the test itself_
> > 
> > ## Does this pull request potentially affect one of the following parts:
> > 
> > * Dependencies (does it add or upgrade a dependency): no
> > * The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
> > * The serializers: no
> > * The runtime per-record code paths (performance sensitive): no
> > * Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
> > * The S3 file system connector: no
> > 
> > ## Documentation
> > 
> > * Does this pull request introduce a new feature? no
> > * If yes, how is the feature documented? not applicable
> > 
> > Here is the script for running the non-dex
> > ```
> > mvn -pl flink-runtime -DnondexRuns=10 edu.illinois:nondex-maven-plugin:2.1.7-SNAPSHOT:nondex -Dtest=org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreSync```
> > ```
> 
> The fix looks okay. One suggestion would be to link you account to the commit. If you check your commit, it does not link to your git account. When you raise a real PR , if your commit is not linked to your git, you wont appear as a contributor.

Resolved! Thanks for reminding me

**Comment by MyEnthusiastic** [comment]
> Nice work on the code changes! Minor comment on the PR description - can you add a short description of the problem and the cause of flakiness in the tests and how the code change addresses the flakiness.

Resolved! make it more clear, included what was the case before change and what change can bring

**Comment by zzjas** [comment]
Just noticed, if you haven't opened a real PR yet, you would want a more descriptive title as well.

**Comment by MyEnthusiastic** [comment]
This PR has already been opened : https://github.com/apache/flink/pull/23464 


And the logic should be same for the test:

- org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreSync
- org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreAsync

**Pull request merged**